### DISCUSSION TITLE
intrinsic parameters of camera heads

### DISCUSSION THREAD

**Discussion by heruoyu1206** [ifm/ifm3d#452] [Q&A]
How to get the intrinsic parameters of the camera heads of O3R, O3D and O3X via Python API and SDK？ Thanks!

**Comment by lola-masson** [comment]
Hi @heruoyu1206,
For the O3R, you can use the TOF_INFO buffer and deserialize it like shown in this example: https://github.com/ifm/ifm3d-examples/blob/main/ovp8xx/python/ovp8xxexamples/core/deserialize_tof_info.py
The TOF_INFO buffer contains the info listed in the API docs: https://api.ifm3d.com/stable/_autosummary/ifm3dpy.deserialize.TOFInfoV4.html, so you should be able to access the intrinsics with `tof_info.instrinsic_calibration` (like is done in the example with `tof_info.illu_temperature` for instance).
For the O3D and O3X I need to check and will get back to you. It should be similar though, with the https://api.ifm3d.com/stable/_autosummary/ifm3dpy.deserialize.O3DInstrinsicCalibration.html

**Reply by heruoyu1206** [reply]
@lola-masson thanks a lot for your quick reply! some following questions like below:
1, How to parse the data array returned by tof_info?  
for the ToF module in O3R camera head, I can get the intrinsic parameters in an array with a length of 32, with the codes: tof_info.intrinsic_calibration.parameters

below is a typical result：
[128.82728576660156, 128.82728576660156, 108.5394058227539, 83.05159759521484, 0.0, 0.0249439999461174, 0.009442999958992004, 0.014127999544143677, 0.0, 1.2100000381469727, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

So, which parameter does each number stands for? Which data is f_x, f_y, which one is cy, k1, p1 and so on?


2, Similar question as Q1, but for RGB module.
got the intrinsic parameters via deserialize_rgb.py, and need to parse the numbers. 

3, Similar question as Q1, but for inverse_intrinsic_calibration.

4, Other than Python, is there any other way to get the intrinsic parameters? if yes, please give some instructions on it. 

thank you!

**Comment by desengph** [comment]
Hello @heruoyu1206 
for further details, please see the documentation on calibration models and parameters here:

https://ifm3d.com/latest/CalibrationRoutines/IntroToCalibrations/README.html

**Comment by heruoyu1206** [comment]
hello, 
Unfortunately, there is no answer in the webpage you mentioned above. 
You know, for the intrinsic parameter, without distortion correction factors, there should be only 4 numbers, which are fx, fy, cx, cy. 
fx is the focal length of the lens in x direction, and fy in y direction. 
cx and cy are the pixel index for the center pixel, determined by the lens center. 

however, in the intrinsic results we acquired via the python API, which could be 

[128.82728576660156, 128.82728576660156, 108.5394058227539, 83.05159759521484, 0.0, 0.0249439999461174, 0.009442999958992004, 0.014127999544143677, 0.0, 1.2100000381469727, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
there are obviously more than 4 numbers provided. so, the distortion parameters must be included. 
i assume the first 4 number stand for the fx, fy, cx, and cy respectively. am I right? 
While, the following number should be distortion (correction) factors, including the radial (k1, k2, k3) and tangential (p1, p2, ...) distortion factors. My question is, what is, which number stands for each factor? 
for example, if someone would like to fuse the amplitude data and the point cloud image, the distortion factors must be clarified. 
thank you!

**Comment by dekhanra** [comment]
Hello @heruoyu1206,

Please see this ![code snippet](https://github.com/ifm/ifm3d-examples/blob/9c4182049fa59652d200e8926f84ebec9c0e59ee/o3d3xx-o3x1xx/DEPRECATED_intrinsics_to_cartesian/python/intrinsics_to_cartesian.ipynb) for the further information

**Reply by heruoyu1206** [reply]
The O3D303 camera model is based on the model described in this document: http://www.vision.caltech.edu/bouguetj/calib_doc/htmls/parameters.html

the doc webpage does not exist. 
Where else can i get this doc please?

**Reply by dekhanra** [reply]
Hello @heruoyu1206,

As the state of code says DEPRECATED, we are no longer maintaining the respective code snippet. But the file you are looking for can be found https://data.caltech.edu/records/jx9cx-fdh55

Download the https://data.caltech.edu/records/jx9cx-fdh55/files/calib_doc.zip?download=1 and look through the index.html page for the further documentation.

Hope this helps.

Cheers!

**Comment by lola-masson** [comment]
Hi @heruoyu1206,

Sorry for the confusion and the deprecated links.

Questions 1, 2 and 2:  
We use two different models, the fish eye model and the Bouguet model. The formulas are both documented here:
https://ifm3d.com/latest/Technology/3D/IntrinsicCalib/intrinsic_calib_3d.html
I just noticed that we did not list the model ID corresponding to the Bouguet model, but it is `modelId=0` for the intrinsics and `modelId=1` for the inverse intrinsics.

When you get the parameters, they will be listed in the order f_x, f_y, m_x, m_y, alpha, k_1, k_2, k_3, k_4, theta_max.

For the RGB camera, you can do the same thing and check the model ID and take the corresponding formula in the document I linked above.

All the info above are for the O3R camera. For the O3D and O3X cameras, some adjustments might be needed.


Question 4:  
This information is only available via ifm3d (Python or C++), in the respective buffers TOF_INFO and RGB_INFO. The Vision Assistant does not currently provide intrinsic calibration information.

Hope this helps.
